### ISSUE TITLE
Better defaults for `exponentiate`

### ISSUE THREAD

**Issue by emstoudenmire** [mtfishman/ITensorTDVP.jl#3]
In the current version, setting exponentiate_krylovdim has a large effect on the speed. However, setting it too low can lead to a sudden failure of accuracy. So it is a tricky thing for a user to have to adjust.

Perhaps the current settings are leading to `exponentiate` use nearly as many Krylov vectors as allowed, possibly because of the `tol` parameter being set to 1E-14. Maybe there is another combination of default settings that will guarantee good accuracy while using the minimal Krylov basis size.

**Comment by emstoudenmire** [comment]
As a related point, it seems the TDVP code currently takes about as long to implement large time steps as smaller time steps. Shouldn't one expect that if the time step is, say, 1.0 that it would take longer and if it was only 0.1 it would take less time, because `exponentiate` wouldn't need to make as many Krylov vectors to accurately do the integration? 

Maybe this is happening, but it would be good to do more precise counts of how many Krylov vectors get made, say for the middle site on the second half sweep. 

If it's fixable, it may be a matter of inputting better parameters to `exponentiate`.

**Comment by mtfishman** [comment]
> As a related point, it seems the TDVP code currently takes about as long to implement large time steps as smaller time steps. Shouldn't one expect that if the time step is, say, 1.0 that it would take longer and if it was only 0.1 it would take less time, because `exponentiate` wouldn't need to make as many Krylov vectors to accurately do the integration?
> 
> Maybe this is happening, but it would be good to do more precise counts of how many Krylov vectors get made, say for the middle site on the second half sweep.
> 
> If it's fixable, it may be a matter of inputting better parameters to `exponentiate`.

Is this all for a fixed accuracy? It seems like the proper comparison would be to evolve for the same time with each timestep to a final state of the same accuracy. Agreed it should be faster with a larger time step. The speedup may also depend on the gap of the system (in general Krylov methods help more when the gap is smaller, for example when compared to power methods).

**Comment by mtfishman** [comment]
> In the current version, setting exponentiate_krylovdim has a large effect on the speed. However, setting it too low can lead to a sudden failure of accuracy. So it is a tricky thing for a user to have to adjust.
> 
> Perhaps the current settings are leading to `exponentiate` use nearly as many Krylov vectors as allowed, possibly because of the `tol` parameter being set to 1E-14. Maybe there is another combination of default settings that will guarantee good accuracy while using the minimal Krylov basis size.

Agreed, the initial implementation just involved copying the DMRG code from ITensors.jl and making minimal modifications, and the defaults for TDVP vs. DMRG for the Krylov solver would be quite different. I think it is best to set a relatively large default for the `krylovdim` (maybe 20 or 30) and then choose a reasonable `tol` like `1e-8`, and have people mostly control the accuracy by adjusting the `tol`.

In fact, for VUMPS/infinite TDVP in `ITensorInfiniteMPS.jl` we use the "precision error" (which is actually the norm of the gradient) as a way to automate determining the Krylov solver tolerance. For example, a good default is setting it to the gradient norm divided by a constant. In practice, using `tol = norm(gradient) / 10` works well off criticality and `tol = norm(gradient) / 100` near criticality. We could consider using a similar approach for finite TDVP.

**Comment by b-kloss** [comment]
I think `tol` in exponentiate here has an intuitive meaning: It's the error made per unit time in solving exp(Ht). It's not related to the gradient norm, and I think `tol` should be a user-defined parameter, with a reasonable default like `1e-8` (to be tested at some point).

**Comment by emstoudenmire** [comment]
Yes, I agree that is likely the thing the user should really be adjusting. I was trying to decide in my head whether it should be that or the `tol=eps/abs(t)` idea mentioned in the KrylovKit docs, but that doesn't seem right within this setting, because in time evolution I believe one does want a certain accuracy per time step, not some total target accuracy at the very end of the whole algorithm.
